src/pcache.cpp

```diff
@@ -0,0 +1,260 @@
+    if(!cstatus || rct == NULL){
+        return false;
+    }
+    reply = static_cast<redisReply*>(redisCommand(rct, "PING"));
+    if(checkReply(reply)){
+        return false;
+    }
+    freeReplyObject(reply);
+    return true;
+}
+
+bool PCache::insert(const std::string& key, const std::string& serializedGeom){
+    //PRINTDB("PCacheReader R: %s = %s", key.c_str() % serializedGeom.c_str() );
+    reply = static_cast<redisReply*>(redisCommand(rct, "SET %s %s", key.c_str(), serializedGeom.c_str()));
+    if(checkReply(reply)){
+        return false;
+    }
+    freeReplyObject(reply);
+    return true;
+}
+
+bool PCache::get(const std::string &key, std::string &serializedGeom){
+    reply = static_cast<redisReply*>(redisCommand(rct, "GET %s", key.c_str()));
+    if(checkReply(reply)){
+        return false;
+    }
+    serializedGeom = reply->str;
+    freeReplyObject(reply);
+    return true;
+}
+// Insert methods takes in key and unserialized Geometry. need to change the function arguments.
+// Serialize the geometry and insert that into redis using insert method
+// Returns on operation success
+bool PCache::insertCGAL(const std::string &key, const shared_ptr<const CGAL_Nef_polyhedron> &N){
+    PRINTDB("Insert: %s", key.c_str());
+    std::stringstream ss;
+    if(N->p3 == nullptr){
+        return insert("CGAL-"+key, "");
+    }
+    ss << *N->p3;
+    std::string data = ss.str();
+    CGAL_cache_entry ce(data);
+    ss.clear();
+    std::stringstream ss1;
+    boost::archive::text_oarchive oa(ss1);
+    oa << ce;
+    return insert("CGAL-"+key, ss1.str());
+}
+
+bool PCache::insertGeometry(const std::string &key, const shared_ptr<const Geometry> &geom){
+    if(geom==nullptr){
+        return insert("GEOM-"+key, "");
+    }
+    if(geom->serializable()){
+        PRINTDB("Insert: %s", key.c_str());
+        std::stringstream ss;
+        Geom_cache_entry ce(geom);
+        boost::archive::text_oarchive oa(ss);
+        oa << ce;
+        std::string data = ss.str();
+        return insert("GEOM-"+key, data);
+    } else return true;
+}
+
+shared_ptr<const CGAL_Nef_polyhedron> PCache::getCGAL(const std::string &key){
+    PRINTDB("Get: %s", key.c_str());
+    std::string data;
+    if(get("CGAL-"+key, data) && !data.empty()){
+        shared_ptr<CGAL_Nef_polyhedron3> p3(new CGAL_Nef_polyhedron3());
+        CGAL_cache_entry ce;
+        std::stringstream ss(data);
+        boost::archive::text_iarchive io(ss);
+        io >> ce;
+        std::string n =ce.N;
+        std::stringstream ss1(n);
+        ss1 >> *p3;
+        shared_ptr<const CGAL_Nef_polyhedron> N(new CGAL_Nef_polyhedron(p3));
+        return N;
+    }
+    shared_ptr<const CGAL_Nef_polyhedron> N(new CGAL_Nef_polyhedron());
+    return N;
+}
+
+shared_ptr<const Geometry> PCache::getGeometry(const std::string &key){
+    PRINTDB("Get: %s", key.c_str());
+    std::string data;
+    shared_ptr<const Geometry> geom;
+    Geom_cache_entry ce;
+    if(get("GEOM-"+key, data)){
```

Comment:
It looks like this condition missing `&& !data.empty()`. I'm getting following error when trying to use persistent cache:
```
terminate called after throwing an instance of 'boost::archive::archive_exception'
  what():  input stream error
```
`PCache::getCGAL` [contains](https://github.com/openscad/openscad/blob/b0ceaad03a13ce089112be72acda3eac5214adec/src/pcache.cpp#L139) such check. Same applies to LCache: `LCache::getCGAL` contains check but `LCache::getGeometry` does not.

Reply:
Thanks, added to the imported branch.